src/charm.py

```diff
@@ -0,0 +1,257 @@
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'tree': '\n'.join(stdout.lines)})
+            case 'ls-1':
+                logger.debug('_on_library_info: tree: executing')
+                process = container.exec(
+                    ['ls', '-1'],
+                    working_dir='/books/',
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'ls-1': '\n'.join(stdout.lines)})
+            case _:
+                msg = (
+                    f'Invalid value {format} for {LIBRARY_INFO_FORMAT_PARAM} parameter'
+                    f' of {LIBRARY_INFO_ACTION} action.'
+                )
+                logger.error(f'_on_library_info: {format}: {msg}')
+                event.fail(msg)
+
+    @staticmethod
+    def get_pebble_layer() -> CalibreWebLayerDict:
+        """Return a dictionary representing the Pebble layer."""
+        c = ' && '.join(
+            [
+                'bash /etc/s6-overlay/s6-rc.d/init-calibre-web-config/run',
+                # with bash because the run script shebang depends on s6
+                'export CALIBRE_DBPATH=/config',
+                'cd /app/calibre-web',
+                'python3 /app/calibre-web/cps.py',
+            ]
+        )
+        command = f"bash -c '{c}'"
+        return {
+            'summary': f'{SERVICE_NAME} layer',
+            'description': f'pebble config layer for {SERVICE_NAME}',
+            'services': {
+                SERVICE_NAME: {
+                    'override': 'replace',
+                    'summary': 'calibre-web',
+                    'command': command,
+                    'startup': 'enabled',
+                    'environment': {
+                        'PUID': '1000',  # copied from example docker run
+                        'PGID': '1000',  # copied from example docker run
+                        'TZ': 'Etc/UTC',  # copied from example docker run
+                    },
+                }
+            },
+        }
+
+    def _push_library_to_storage(self) -> None:
+        """Push user provided or default calibre-library resource to storage."""
+        library_write_behaviour, _ = self._get_library_write_behaviour()
+        if library_write_behaviour is None:
+            return
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
+        if contents := container.list_files('/books/'):
+            msg_prefix = f'_push_library_to_storage: {library_write_behaviour=}: '
+            match library_write_behaviour:
+                case 'skip':
+                    logger.debug(msg_prefix + 'returning')
+                    return
+                case 'clean':
+                    logger.debug(msg_prefix + 'cleaning ...')
+                    for fileinfo in contents:
```

Comment:
Rather than iterating through each item in `/books/` removing them one by one, couldn't you just remove `/books/` (and then create an empty replacement) instead? It could be significantly fewer Pebble calls with a large library, I think.

Reply:
This would be ideal, except that `/books/` is actually a mounted volume, so it can't be removed and recreated so easily.

Reply:
Ah, right. If you continue with the charm, it might be worth putting the books folder into a folder in the mount so you can get this shortcut.

Might also be interesting reading this (closed, not complete) issue: https://github.com/canonical/operator/issues/1042

Reply:
Thank you, helpful content in that issue. I think if I figure out how to configure the default library location in calibre-web automatically, bypassing the need for the user to write '/books/subfolder', then I'd just do the subfolder thing.